### PR TITLE
Update from original

### PR DESCRIPTION
**Pull request by DmitryBochkarev**
Добавится with_hold
https://github.com/abak-press/apress-company_abouts/pull/35/files#r69853069

### CODE REVIEW

File: test/test_postgresql_cursor.rb [hunk 1]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+      assert r.has_key?("id")
+      r = cursor.fetch(symbolize_keys:true)
+      assert r.has_key?(:id)
+      cursor.close
+    end
+  end
+
+  def test_bad_sql
+    begin
+      ActiveRecord::Base.each_row_by_sql('select * from bad_table') { }
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:115]: Space inside empty braces detected.

File: test/test_postgresql_cursor.rb [hunk 2]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+    assert_equal 3, items
+  end
+
+  def test_fetch_symbolize_keys
+    Product.transaction do
+      #cursor = PostgreSQLCursor::Cursor.new("select * from products order by 1")
+      cursor = Product.all.each_row
+      r = cursor.fetch
+      assert r.has_key?("id")
+      r = cursor.fetch(symbolize_keys:true)
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:107]: Space missing after colon.

File: test/test_postgresql_cursor.rb [hunk 3]
```diff
@@ -78,5 +86,36 @@ def test_pluck
     assert_equal 1000, r.size
     assert_equal Fixnum, r.first.class
   end
-  
+
+  def test_with_hold
+    items = 0
+    Product.where("id < 4") .each_instance(with_hold: true, block_size:1) do |row|
+      Product.transaction do
+        row.update(data:Time.now.to_f.to_s)
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:94]: Space missing after colon.

File: test/test_postgresql_cursor.rb [hunk 4]
```diff
@@ -78,5 +86,36 @@ def test_pluck
     assert_equal 1000, r.size
     assert_equal Fixnum, r.first.class
   end
-  
+
+  def test_with_hold
+    items = 0
+    Product.where("id < 4") .each_instance(with_hold: true, block_size:1) do |row|
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:92]: Space missing after colon.

File: test/test_postgresql_cursor.rb [hunk 5]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+      r = cursor.fetch
+      assert r.has_key?("id")
+      r = cursor.fetch(symbolize_keys:true)
+      assert r.has_key?(:id)
+      cursor.close
+    end
+  end
+
+  def test_bad_sql
+    begin
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:114]: Redundant `begin` block detected.

File: test/test_postgresql_cursor.rb [hunk 6]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+        row.update(data:Time.now.to_f.to_s)
+        items += 1
+      end
+    end
+    assert_equal 3, items
+  end
+
+  def test_fetch_symbolize_keys
+    Product.transaction do
+      #cursor = PostgreSQLCursor::Cursor.new("select * from products order by 1")
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:103]: Missing space after #.

File: test/test_postgresql_cursor.rb [hunk 7]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+      cursor = Product.all.each_row
+      r = cursor.fetch
+      assert r.has_key?("id")
+      r = cursor.fetch(symbolize_keys:true)
+      assert r.has_key?(:id)
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:108]: `Hash#has_key?` is deprecated in favor of `Hash#key?`.

File: test/test_postgresql_cursor.rb [hunk 8]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+    end
+    assert_equal 3, items
+  end
+
+  def test_fetch_symbolize_keys
+    Product.transaction do
+      #cursor = PostgreSQLCursor::Cursor.new("select * from products order by 1")
+      cursor = Product.all.each_row
+      r = cursor.fetch
+      assert r.has_key?("id")
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:106]: `Hash#has_key?` is deprecated in favor of `Hash#key?`.

File: test/test_postgresql_cursor.rb [hunk 9]
```diff
@@ -78,5 +86,36 @@ def test_pluck
...
+      assert r.has_key?(:id)
+      cursor.close
+    end
+  end
+
+  def test_bad_sql
+    begin
+      ActiveRecord::Base.each_row_by_sql('select * from bad_table') { }
+      raise "Did Not Raise Expected Exception"
+    rescue Exception => e
```
Comment (apress-hound) [inline test/test_postgresql_cursor.rb:117]: Avoid rescuing the `Exception` class. Perhaps you meant to rescue `StandardError`?

File: lib/postgresql_cursor/cursor.rb [hunk 10]
```diff
@@ -160,23 +179,32 @@ def column_types
...
     # Private: Fetches the next block of rows into @block
     def fetch_block(block_size=nil)
       block_size ||= @block_size ||= @options.fetch(:block_size) { 1000 }
       @result = @connection.execute("fetch #{block_size} from cursor_#{@cursor}")
-      @block = @result.collect {|row| row } # Make our own
+
+      if @iterate == :each_array
+        @block = @result.each_row.collect {|row| row }
+      else
+        @block = @result.collect {|row| row }
```
Comment (apress-hound) [inline lib/postgresql_cursor/cursor.rb:206]: Space between { and | missing.

**Comment by DmitryBochkarev** [comment]
@bibendi @Napolskih

**Comment by c1n1c** [comment]
так там же чувак принял реквест с даунгрейдом до 3.1, мы же можем 0.6 просто подключать

**Pull request closed without merging**